### PR TITLE
process-util: don't use overly large buffer to store process command line

### LINKED ISSUE #11502: systemd-journald: very large virt memory leak (heap)
Debian 240-4

VSS/VSIZE growing very rapidly without bound.

https://github.com/systemd/systemd/issues/9141?fbclid=IwAR0ji5RLdD6bxZH0erUdl2IUymVWvrONdWmUYm1qgfvK2ZyMX5SVlOMVE3U#issuecomment-454126686

https://bugzilla.redhat.com/show_bug.cgi?id=1665931
```
$ sudo pmap --extended $(pidof systemd-journald) | sort -rnk 2 | head
0000563884697000 7910112   33504   33504 rw---   [ anon ]
00007f6950a81000    8192    7432    7432 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f6950080000    8192    6992    6992 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f694f880000    8192    7272    7272 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f694f080000    8192    7388    7388 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f694e880000    8192    6616    6616 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f694e080000    8192    6896    6896 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
00007f694d0b5000    8192    3368    3368 rw-s- system.journal
00007f694c8b5000    8192       8       8 rw---   [ anon ]
00007f694d8b5000    7980    4308    4308 rw-s- system@2bbaf66c15df446c9f33ec3169b8d32a-0000000000000001-00057fc6d38900f9.journal
```
-XX says its heap
```
$ sudo pmap -XX $(pidof systemd-journald)
9265:   /lib/systemd/systemd-journald
         Address Perm   Offset Device   Inode    Size KernelPageSize MMUPageSize   Rss   Pss Shared_Clean Shared_Dirty Private_Clean Private_Dirty Referenced Anonymous LazyFree AnonHugePages ShmemPmdMapped Shared_Hugetlb Private_Hugetlb Swap SwapPss Locked              VmFlags Mapping
....
    563884697000 rw-p 00000000  00:00       0 7946976              4           4 33652 33652            0            0             0         33652      30888     33652        0             0              0              0               0    0       0      0    rd wr mr mw me ac [heap]
```
Grows without bound until it is restarted
```
~$ sudo ps axuk-vsz | head
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      9265  0.0  1.0 8005204 89416 ?       Ss   Jan18   0:09 /lib/systemd/systemd-journald
mysql     1025  0.3  2.5 1867820 207312 ?      Sl   Jan18   5:47 /usr/sbin/mysqld --daemonize --pid-file=/run/mysqld/mysqld.pid
nyet       987  0.2  0.2 1016720 23752 ?       Ssl  Jan18   4:14 /usr/bin/rslsync --config /etc/resilio-sync/nyet.conf
root       938  0.2  0.3 848372 24728 ?        Ssl  Jan18   3:59 /usr/bin/python3 /usr/bin/fail2ban-server -xf start
clamav     710  0.0  7.4 842544 604508 ?       Ssl  Jan18   1:04 /usr/sbin/clamd --foreground=true
bind       966  0.0  0.7 446904 60704 ?        Ssl  Jan18   0:46 /usr/sbin/named -u bind
nobody     912  0.0  0.0 409888  3296 ?        Ssl  Jan18   0:19 /usr/bin/memcached -m 64 -p 11211 -u nobody -l 127.0.0.1
root       699  0.0  0.1 403960 11484 ?        Ssl  Jan18   0:00 /usr/lib/udisks2/udisksd
_graphi+ 16835  0.0  0.5 389808 45980 ?        Sl   00:00   0:03 (wsgi:_graphite)  -k start
```
More importantly, you can see it was fine until I "upgraded" systemd on around Jan 14
```
2019-01-14 06:26:04 upgrade systemd:amd64 240-3 240-4
```
memory-pinpoint 1547269562 1547960762

### PR DESCRIPTION
**Pull request by msekletar**
Allocate new string as a return value and free our "scratch pad"
buffer that is potentially much larger than needed (up to
_SC_ARG_MAX).

Fixes #11502

### CODE REVIEW

File: src/basic/process-util.c [hunk 1]
```diff
@@ -243,7 +243,15 @@ int get_process_cmdline(pid_t pid, size_t max_length, bool comm_fallback, char *
...
         }
 
-        *line = ans;
+        k = strdup(ans);
+        if (!k) {
+                free(ans);
+                return -ENOMEM;
+        }
+
+        free(ans);
```
Comment (keszybz) [inline src/basic/process-util.c]: Please use `_cleanup_free_` instead. It'll be necessary to add `TAKE_PTR` in a few places.

**Comment by msekletar** [comment]
> I don't think this is a the whole solution. It seems the "leak" is growing and growing, so just making smaller allocations cannot be the whole answer, but this looks like a step in the right direction. So let's merge this, as it should make things much nicer, and look for further improvements later.

Maybe we should track how big is each ```ClientContext``` entry and have limits on how much memory overall can cache consume.

**Comment by msekletar** [comment]
@keszybz I've updated my branch. PTAL.

**Comment by msekletar** [comment]
Ubuntu CI failures are most likely unrelated and seem to be due to packaging issues,

```
E: There were unauthenticated packages and -y was used without --allow-unauthenticated
autopkgtest: WARNING: Test dependencies are unsatisfiable - calling apt install on test deps directly for further data about failing dependencies in test logs
boot-smoke           FAIL badpkg
```

**Comment by poettering** [comment]
The allocation of the sysconf size is still in there. The upper loop should be reworked to do GREEDY_REALLOC() really, so that the buffer increases exponentially in size as needed.

**Comment by keszybz** [comment]
> GREEDY_REALLOC()

I thought about this, but there's an argument for doing this as in the current patch: those strings will be kept in a cache, potentially for a long time, so it is quite worthwhile to minimize the allocation. And if we are going to do the `strdup` at the end anyway, and the maximum size is fixed anyway and not that large, then it's OK to not use greedy_realloc.

**Comment by poettering** [comment]
> > GREEDY_REALLOC()
> 
> I thought about this, but there's an argument for doing this as in the current patch: those strings will be kept in a cache, potentially for a long time, so it is quite worthwhile to minimize the allocation. And if we are going to do the `strdup` at the end anyway, and the maximum size is fixed anyway and not that large, then it's OK to not use greedy_realloc.

it appears to me the right approach is to use greedy_realloc() initially, and then use realloc() at the end to give up the space not actually needed. realloc() is pretty ok for giving up unneeded memory, much better i think than copying yourself

**Pull request merged**